In [1]:
import pandas as pd
import os

est = "C:\Users\wb474294\Box Sync\DIME\Rio De Janeiro\RAIS_dropbox\Estb2014\Estb2014.TXT"
municipal_code = "C:\Users\wb474294\Box Sync\DIME\Rio De Janeiro\Municipio Codes.txt"

### Extract the following variables from Estb2014 data
+ Municpio: municipality code (use Municipio_Codes.csv to retrieve the names)
+ Nomelogradouro: street name 
+ Nmerologradouro: number 
+ Nomebairro: neighboorhood 
+ CEP Estab: zip code

In [2]:
# Read data
df_est = pd.read_csv(est,sep=';',encoding = 'latin1')
df_mc = pd.read_csv(municipal_code,sep='\t',encoding = 'latin1')

In [3]:
# print df.columns
print df_est.iloc[:10, [15,16,17,13,1]] 

                            Nome Logradouro  Número Logradouro  \
0                            MAJOR AMARANTE               3160   
1                     RUA JOSE VIEIRA CAULA               5651   
2    GETULIO VARGAS C AV TANCREDO DE A NEVE                  1   
3                               CASCALHEIRA                283   
4       R PRESBITERO HONORATO PEREIRA, 1831               1831   
5                                  GUANAMBI               1872   
6                 RODOVIA  RO -492   KM 101                101   
7                   AVENIDA DUQUE DE CAXIAS               2211   
8                            RIO DE JANEIRO               1998   
9                       AV TRANSCONTINENTAL               1559   

                      Nome Bairro  Município  CEP Estab  
0                          CENTRO     110030   76980000  
1                         IGARAPE     110020   76824335  
2                          CENTRO     110037   76952000  
3                      VILA JOTAO     110

In [5]:
print df_mc.columns
print df_mc.iloc[:10, [0,1,2,3,7,8,9]]

Index([u'UF', u'Name_UF', u'Mesorregião Geográfica', u'Nome_Mesorregião',
       u'Microrregião Geográfica', u'Nome_Microrregião', u'Município',
       u'Municipio Code', u'Municipio Code 6-digit', u'Name_Município'],
      dtype='object')
   UF   Name_UF  Mesorregião Geográfica   Nome_Mesorregião  Municipio Code  \
0  11  Rondônia                       2  Leste Rondoniense         1100015   
1  11  Rondônia                       2  Leste Rondoniense         1100023   
2  11  Rondônia                       2  Leste Rondoniense         1100031   
3  11  Rondônia                       2  Leste Rondoniense         1100049   
4  11  Rondônia                       2  Leste Rondoniense         1100056   
5  11  Rondônia                       2  Leste Rondoniense         1100064   
6  11  Rondônia                       2  Leste Rondoniense         1100072   
7  11  Rondônia                       1    Madeira-Guaporé         1100080   
8  11  Rondônia                       2  Leste Rondoniense

In [11]:
# Extract the relevant columns only
df_address = df_est.iloc[:, [15,16,17,13,1]] 
df_municipio = df_mc.iloc[:, [0,1,2,3,7,8,9]][df_mc.UF==33] #selecting State of Rio only 
print df_address.shape, df_municipio.shape

df_address.columns = ["street_name","street_number","neighborhood","muni_code6","zipcode"]
df_municipio.columns = ["state_code","state","region_code","region","muni_code","muni_code6","municipal"]
df_municipio

(8240846, 5) (92, 7)


state_code           state  region_code  \
3175          33  Rio de Janeiro            5   
3176          33  Rio de Janeiro            1   
3177          33  Rio de Janeiro            4   
3178          33  Rio de Janeiro            3   
3179          33  Rio de Janeiro            4   
3180          33  Rio de Janeiro            4   
3181          33  Rio de Janeiro            5   
3182          33  Rio de Janeiro            5   
3183          33  Rio de Janeiro            6   
3184          33  Rio de Janeiro            3   
3185          33  Rio de Janeiro            1   
3186          33  Rio de Janeiro            4   
3187          33  Rio de Janeiro            6   
3188          33  Rio de Janeiro            1   
3189          33  Rio de Janeiro            2   
3190          33  Rio de Janeiro            3   
3191          33  Rio de Janeiro            2   
3192          33  Rio de Janeiro            3   
3193          33  Rio de Janeiro            2   
3194          33  Rio de Janeiro            3   
3195          33  Rio de Janeiro            4   
3196          33  Rio de Janeiro            2   
3197          33  Rio de Janeiro            3   
3198          33  Rio de Janeiro            3   
3199          33  Rio de Janeiro            6   
3200          33  Rio de Janeiro            6   
3201          33  Rio de Janeiro            6   
3202          33  Rio de Janeiro            4   
3203          33  Rio de Janeiro            6   
3204          33  Rio de Janeiro            6   
...          ...             ...          ...   
3237          33  Rio de Janeiro            5   
3238          33  Rio de Janeiro            6   
3239          33  Rio de Janeiro            5   
3240          33  Rio de Janeiro            5   
3241          33  Rio de Janeiro            4   
3242          33  Rio de Janeiro            6   
3243          33  Rio de Janeiro            3   
3244          33  Rio de Janeiro            1   
3245          33  Rio de Janeiro            2   
3246          33  Rio de Janeiro            2   
3247          33  Rio de Janeiro            6   
3248          33  Rio de Janeiro            2   
3249          33  Rio de Janeiro            6   
3250          33  Rio de Janeiro            1   
3251          33  Rio de Janeiro            6   
3252          33  Rio de Janeiro            4   
3253          33  Rio de Janeiro            3   
3254          33  Rio de Janeiro            3   
3255          33  Rio de Janeiro            4   
3256          33  Rio de Janeiro            6   
3257          33  Rio de Janeiro            4   
3258          33  Rio de Janeiro            3   
3259          33  Rio de Janeiro            6   
3260          33  Rio de Janeiro            6   
3261          33  Rio de Janeiro            3   
3262          33  Rio de Janeiro            3   
3263          33  Rio de Janeiro            5   
3264          33  Rio de Janeiro            1   
3265          33  Rio de Janeiro            6   
3266          33  Rio de Janeiro            5   

                               region  muni_code  muni_code6  \
3175                   Sul Fluminense    3300100      330010   
3176              Noroeste Fluminense    3300159      330015   
3177                         Baixadas    3300209      330020   
3178                Centro Fluminense    3300225      330022   
3179                         Baixadas    3300233      330023   
3180                         Baixadas    3300258      330025   
3181                   Sul Fluminense    3300308      330030   
3182                   Sul Fluminense    3300407      330040   
3183  Metropolitana do Rio de Janeiro    3300456      330045   
3184                Centro Fluminense    3300506      330050   
3185              Noroeste Fluminense    3300605      330060   
3186                         Baixadas    3300704      330070   
3187  Metropolitana do Rio de Janeiro    3300803      330080   
3188              Noroeste Fluminense    3300902      330090   
3189   

In [12]:
# combine the data together
# df = pd.merge(df_address, df_municipio, how="left", on='municipal_code')
df = pd.merge(df_address, df_municipio, on='muni_code6')
print df.shape
print df[:10]

(561684, 11)
                                street_name  street_number  \
0                           TRAVESSA TALITA            142   
1                         ESTR DA CONCEICAO              9   
2              RUA DOUTOR FRANCISCO PORTELA           2713   
3                JORNALISTA ROBERTO MARINHO            221   
4                JORNALISTA ROBERTO MARINHO            221   
5                        ESTRADA DO PACHECO             20   
6                   AVENIDA FELICIANO SODRE            225   
7                      RUA DOUTOR JURUMENHA           2545   
8              RUA JOSE LOURENCO DE AZEVEDO             81   
9                        AFONSO GOMES DUTRA             60   

                     neighborhood  muni_code6   zipcode  state_code  \
0                     SAO GONCALO      330490  24456320          33   
1                          ITAUNA      330490  24461840          33   
2                       PATRONATO      330490  24435000          33   
3                   

In [40]:
# for col in ["street_number","zipcode"]:
#     df[col] = df[col].astype(str)
#     df[col] = df[col].str.strip()

for i in df.zipcode.unique():
    if len(i) <8:
        print i
df.address[df.zipcode=="1045001"]
        
# for col in ["street_name","neighborhood","municipal","state"]:
#     df[col] = df[col].str.strip()

1045001


77396    PRACA DA REPUBLICA 473, REPUBLICA, Rio de Jane...
Name: address, dtype: object

In [58]:
# df['addr'] = df.street_name.map(str)+" "+df.street_number+", "+df.municipal+" "+df.state_name+", "+df.zipcode
# df['country'] = 'Brazil'
# df['street'] = df.street_name.map(str)+" "+df.street_number

In [38]:
# # The data set is large so devide it into chunks of n obs
# n = 10000
# # x = df.shape[0]//n
# df_export = df[['addr','zipcode']]

# if not os.path.exists("address"):
#     os.makedirs("address")

# for i in xrange(0, df.shape[0]-n, n):
#     df_export.iloc[i:i+n,:].to_csv('address/address%i-%i.txt'%(i,i+n-1), sep=',',encoding = 'utf-8')
# df_export.iloc[i+n:,:].to_csv('address/address%i-%i.txt'%(i+n,df.shape[0]), sep=',',encoding = 'utf-8')

In [53]:
# #extract the address part only
# df_export = df[['addr','zipcode']]
# df_export.to_csv('address.txt', sep=',',encoding = 'latin1')

In [47]:
# len(df['state_name'].unique())
# for x in sorted(df['state_name'].unique()):
#     print x.encode("utf-8")

Acre
Alagoas
Amapá
Amazonas
Bahia
Ceará
Distrito Federal
Espírito Santo
Goiás
Maranhão
Mato Grosso
Mato Grosso do Sul
Minas Gerais
Paraná
Paraíba
Pará
Pernambuco
Piauí
Rio Grande do Norte
Rio Grande do Sul
Rio de Janeiro
Rondônia
Roraima
Santa Catarina
Sergipe
São Paulo
Tocantins


In [28]:
# print len(df.municipal.value_counts()), len(df.neighborhood.value_counts())
# df.municipal.value_counts().to_csv('Rio_Municipals.txt', sep=',',encoding = 'latin1')

df['country'] = 'Brazil'
df['street'] = df.street_name.map(str)+" "+df.street_number+" - "+df.neighborhood
df['address'] = df.street_name.map(str)+" "+df.street_number+" - "+df.neighborhood+", " \
                                                     +df.municipal+" - "+df.state+", "+df.zipcode+", "+df.country

In [22]:
# Function to devide the data into chunks of n obs and save as csv
def save_addr(n, df, s):
    if not os.path.exists(s):
        os.makedirs(s)

    for i in xrange(0, df.shape[0]-n, n):
        df.iloc[i:i+n,:].to_csv('%s/%s%i-%i.txt'%(s,s,i,i+n-1), sep=',',encoding = 'latin1')
    df.iloc[i+n:,:].to_csv('%s/%s%i-%i.txt'%(s,s,i+n,df.shape[0]), sep=',',encoding = 'latin1')
    
# Function to devide the data by municipals and save as csv
def save_addr_by_municipal(df, s):
    if not os.path.exists(s):
        os.makedirs(s)
        
    for m in sorted(df['municipal'].unique()):
        if m =="Rio de Janeiro":
            save_addr(15000,df[df.municipal==m], s)
        else:
            df[df.municipal==m].to_csv('%s/%s.txt'%(s,m), sep=',',encoding='latin1')
        
    
n = 10000
df_export = df[df.region_code==6].iloc[:,[8,2,5,6,4,7]]
df_export.index = df_export.index - 2711124 #reset index to 0
dirname = "Rio_metro_address"
# save_addr(n,df_export,dirname)
save_addr_by_municipal(df_export,dirname)
# df_export.iloc[0:3000,:].to_csv('0-3000_neighbor.txt', sep=',',encoding = 'latin1')

IndexError: positional indexers are out-of-bounds

In [31]:
print df.columns
print df[df.region_code==6].shape
# df.region.value_counts()
print df[:1]
df.address[0].encode("latin1")

Index([u'street_name', u'street_number', u'neighborhood', u'muni_code6',
       u'zipcode', u'state_code', u'state', u'region_code', u'region',
       u'muni_code', u'municipal', u'country', u'street', u'address'],
      dtype='object')
(425741, 14)
       street_name street_number neighborhood  muni_code6   zipcode  \
0  TRAVESSA TALITA           142  SAO GONCALO      330490  24456320   

   state_code           state  region_code                           region  \
0          33  Rio de Janeiro            6  Metropolitana do Rio de Janeiro   

   muni_code    municipal country                             street  \
0    3304904  São Gonçalo  Brazil  TRAVESSA TALITA 142 - SAO GONCALO   

                                             address  
0  TRAVESSA TALITA 142, SAO GONCALO, São Gonçalo,...  


'TRAVESSA TALITA 142, SAO GONCALO, S\xe3o Gon\xe7alo, Rio de Janeiro, 24456320 Brazil'

### Pending issues:
+ Address sometimes do not exist in google so approximating the location by zipcode nessesary. i.e. Row 1 "RUA JOSE VIEIRA CAULA 5651" (street name does not exist) or Row 6 "RODOVIA  RO -492   KM 101" --> "RO-492 101" (wrong representation?)

Test using 0-9999: 203 out of 2500 was not mapped by Google API.  
Neighborhood variable var: 184 out of 2500
whole address thing: 
--> To Do: Try incorporate neighboorhood variable to see if improves mapping. Try one-line address for comparison.

### Create a pipeline such that:
1. first send the whole address
2. if error, record that the address didn't work and then try zip code.